File: tools/liblib_generator.js
/**
 * LiblibAI 图片生成脚本
 * 使用 LiblibAI API 生成游戏美术资源
 */

const crypto = require('crypto');
const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');

// API 配置
const ACCESS_KEY = "r0LTLSLDq4e3a-PFdmrBpA";
const SECRET_KEY = "4lH_K_Ya66ENhg-h_C64YnpnZdUoU6iJ";
const BASE_URL = "openapi.liblibai.cloud";
const TEXT2IMG_URI = "/api/generate/webui/text2img";
const STATUS_URI = "/api/generate/webui/status";

// 输出目录
const OUTPUT_DIR = "E:\\AI_Project\\MS\\MoShou\\Assets\\Resources\\Sprites\\Generated";

// 任务文件
const TASK_FILE = "E:\\AI_Project\\MS\\.task\\art_generation_task.json";

/**
 * 生成 HMAC-SHA1 签名
 */
function generateSignature(uri, timestamp, nonce) {
    const stringToSign = `${uri}&${timestamp}&${nonce}`;
    const hmac = crypto.createHmac('sha1', SECRET_KEY);
    hmac.update(stringToSign);
    const signature = hmac.digest('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');
    return signature;
}

/**
 * 生成随机字符串
 */
function generateNonce(length = 16) {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
        result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
}

/**
 * 发送 API 请求
 */
function makeApiRequest(uri, payload) {
    return new Promise((resolve, reject) => {
        const timestamp = Date.now().toString();
        const nonce = generateNonce();
        const signature = generateSignature(uri, timestamp, nonce);

        const queryParams = new URLSearchParams({
            AccessKey: ACCESS_KEY,
            Signature: signature,
            Timestamp: timestamp,
            SignatureNonce: nonce
        });

        const body = JSON.stringify(payload);

        const options = {
            hostname: BASE_URL,
            port: 443,
            path: `${uri}?${queryParams.toString()}`,
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(body)
            }
        };

        console.log(`[DEBUG] Request URL: https://${BASE_URL}${uri}?${queryParams.toString()}`);

        const req = https.request(options, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(data));
                } catch (e) {
                    console.error('[ERROR] Failed to parse response:', data);
                    reject(e);
                }
            });
        });

        req.on('error', reject);
        req.write(body);
        req.end();
    });
}

/**
 * 检查生成任务状态
 */
function checkStatus(generateUuid) {
    return makeApiRequest(STATUS_URI, { generateUuid });
}

/**
 * 下载图片
 */
function downloadImage(url, filename) {
    return new Promise((resolve, reject) => {
        const filepath = path.join(OUTPUT_DIR, filename);
        const file = fs.createWriteStream(filepath);

        const protocol = url.startsWith('https') ? https : http;
        protocol.get(url, (response) => {
            // 处理重定向
            if (response.statusCode === 301 || response.statusCode === 302) {
                downloadImage(response.headers.location, filename)
                    .then(resolve)
                    .catch(reject);
                return;
            }

            response.pipe(file);
            file.on('finish', () => {
                file.close();
                console.log(`[SUCCESS] Downloaded: ${filename}`);
                resolve(true);
            });
        }).on('error', (err) => {
            fs.unlink(filepath, () => {}); // 删除失败的文件
            console.error(`[ERROR] Download failed: ${err.message}`);
            reject(err);
        });
    });
}

/**
 * 等待指定毫秒
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * 生成单张图片
 */
async function generateImage(prompt, negativePrompt, width, height, filename) {
    console.log(`\n[INFO] Generating: ${filename}`);
    console.log(`  Prompt: ${prompt.substring(0, 50)}...`);

    // 构建请求payload
    const payload = {
        templateUuid: "5d7e67009b344550bc1aa6ccbfa1d7f4",  // 通用文生图模板
        generateParams: {
            prompt: prompt,
            negativePrompt: negativePrompt,
            width: width,
            height: height,
            steps: 20,
            cfgScale: 7,
            seed: -1,
            samplerName: "DPM++ 2M Karras"
        }
    };

    try {
        // 发送生成请求
        const result = await makeApiRequest(TEXT2IMG_URI, payload);
        console.log('[DEBUG] Generate response:', JSON.stringify(result).substring(0, 200));

        if (result.code !== 0) {
            console.error(`[ERROR] Generation failed: ${JSON.stringify(result)}`);
            return false;
        }

        const generateUuid = result.data?.generateUuid;
        if (!generateUuid) {
            console.error('[ERROR] No generateUuid in response');
            return false;
        }

        console.log(`  Task UUID: ${generateUuid}`);

        // 轮询状态
        const maxAttempts = 60;
        for (let attempt = 0; attempt < maxAttempts; attempt++) {
            await sleep(5000); // 每5秒检查一次

            const statusResult = await checkStatus(generateUuid);
            console.log(`[DEBUG] Full status response: ${JSON.stringify(statusResult).substring(0, 500)}`);

            const status = statusResult.data?.generateStatus || statusResult.data?.status;
            const percentCompleted = statusResult.data?.percentCompleted || statusResult.data?.progress || 0;

            console.log(`  Status check ${attempt + 1}/${maxAttempts}: ${status} (${percentCompleted}%)`);

            // generateStatus: 5 = SUCCESS, 4 = PROCESSING, 2 = FAILED
            if (status === 5 || status === 'SUCCESS') {
                const images = statusResult.data?.images || [];
                if (images.length > 0) {
                    const imageUrl = images[0].imageUrl;
                    if (imageUrl) {
                        return await downloadImage(imageUrl, filename);
                    }
                }
                console.error('[ERROR] No image URL in response');
                return false;
            } else if (status === 2 || status === 'FAILED') {
                console.error('[ERROR] Generation failed');
                return false;
            }
        }

        console.error('[ERROR] Timeout waiting for generation');
        return false;

    } catch (error) {
        console.error(`[ERROR] Exception: ${error.message}`);
        return false;
    }
}

/**
 * 主函数
 */
async function main() {
    console.log('='.repeat(60));
    console.log('LiblibAI Art Generator');
    console.log('='.repeat(60));

    // 确保输出目录存在
    if (!fs.existsSync(OUTPUT_DIR)) {
        fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }

    // 读取任务文件
    let taskData;
    try {
        const taskContent = fs.readFileSync(TASK_FILE, 'utf-8');
        taskData = JSON.parse(taskContent);
    } catch (error) {
        console.error(`[ERROR] Failed to read task file: ${error.message}`);
        process.exit(1);
    }

    const requests = taskData.generation_requests || [];
    console.log(`[INFO] Found ${requests.length} generation requests`);
    console.log(`[INFO] Output directory: ${OUTPUT_DIR}`);

    let successCount = 0;
    let failCount = 0;

    // 只处理第一个pending的任务作为测试
    for (const req of requests) {
        if (req.status === 'completed') {
            console.log(`\n[SKIP] Already completed: ${req.filename}`);
            continue;
        }

        const success = await generateImage(
            req.prompt,
            req.negative_prompt || '',
            req.width,
            req.height,
            req.filename
        );

        if (success) {
            req.status = 'completed';
            successCount++;
        } else {
            req.status = 'failed';
            failCount++;
        }

        // 更新任务文件
        fs.writeFileSync(TASK_FILE, JSON.stringify(taskData, null, 2), 'utf-8');

        // 为避免API限流，等待一下
        console.log('\n[INFO] Waiting 3 seconds before next request...');
        await sleep(3000);
    }

    console.log(`\n${'='.repeat(60)}`);
    console.log(`[DONE] Success: ${successCount}, Failed: ${failCount}`);
}

main().catch(console.error);

File: tools/liblib_generator.py
#!/usr/bin/env python3
"""
LiblibAI 图片生成脚本
使用 LiblibAI API 生成游戏美术资源
"""

import hmac
import hashlib
import base64
import time
import random
import string
import json
import requests
import os
from urllib.parse import urlencode

# API 配置
ACCESS_KEY = "r0LTLSLDq4e3a-PFdmrBpA"
SECRET_KEY = "4lH_K_Ya66ENhg-h_C64YnpnZdUoU6iJ"
BASE_URL = "https://openapi.liblibai.cloud"
TEXT2IMG_URI = "/api/generate/webui/text2img"
STATUS_URI = "/api/generate/webui/status"

# 输出目录
OUTPUT_DIR = r"E:\AI_Project\MS\MoShou\Assets\Resources\Sprites\Generated"

def generate_signature(uri, timestamp, nonce):
    """生成 HMAC-SHA1 签名"""
    # 拼接待签名字符串
    string_to_sign = f"{uri}&{timestamp}&{nonce}"

    # HMAC-SHA1 加密
    signature = hmac.new(
        SECRET_KEY.encode('utf-8'),
        string_to_sign.encode('utf-8'),
        hashlib.sha1
    ).digest()

    # Base64 URL-safe 编码并移除尾部等号
    signature_b64 = base64.urlsafe_b64encode(signature).decode('utf-8').rstrip('=')

    return signature_b64

def generate_nonce(length=16):
    """生成随机字符串"""
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))

def make_api_request(uri, payload):
    """发送 API 请求"""
    timestamp = str(int(time.time() * 1000))
    nonce = generate_nonce()
    signature = generate_signature(uri, timestamp, nonce)

    # 构建查询参数
    query_params = {
        "AccessKey": ACCESS_KEY,
        "Signature": signature,
        "Timestamp": timestamp,
        "SignatureNonce": nonce
    }

    url = f"{BASE_URL}{uri}?{urlencode(query_params)}"

    headers = {
        "Content-Type": "application/json"
    }

    response = requests.post(url, json=payload, headers=headers)
    return response.json()

def check_status(generate_uuid):
    """检查生成任务状态"""
    timestamp = str(int(time.time() * 1000))
    nonce = generate_nonce()
    signature = generate_signature(STATUS_URI, timestamp, nonce)

    query_params = {
        "AccessKey": ACCESS_KEY,
        "Signature": signature,
        "Timestamp": timestamp,
        "SignatureNonce": nonce
    }

    url = f"{BASE_URL}{STATUS_URI}?{urlencode(query_params)}"

    payload = {
        "generateUuid": generate_uuid
    }

    response = requests.post(url, json=payload, headers={"Content-Type": "application/json"})
    return response.json()

def download_image(url, filename):
    """下载图片到本地"""
    response = requests.get(url)
    if response.status_code == 200:
        filepath = os.path.join(OUTPUT_DIR, filename)
        with open(filepath, 'wb') as f:
            f.write(response.content)
        print(f"[SUCCESS] Downloaded: {filename}")
        return True
    else:
        print(f"[ERROR] Failed to download: {filename}")
        return False

def generate_image(prompt, negative_prompt, width, height, filename):
    """生成单张图片"""
    print(f"\n[INFO] Generating: {filename}")
    print(f"  Prompt: {prompt[:50]}...")

    # 构建请求payload - 使用通用模板
    payload = {
        "templateUuid": "5d7e67009b344550bc1aa6ccbfa1d7f4",  # 通用文生图模板
        "generateParams": {
            "prompt": prompt,
            "negativePrompt": negative_prompt,
            "width": width,
            "height": height,
            "steps": 20,
            "cfgScale": 7,
            "seed": -1,  # 随机种子
            "samplerName": "DPM++ 2M Karras"
        }
    }

    # 发送生成请求
    result = make_api_request(TEXT2IMG_URI, payload)

    if result.get("code") != 0:
        print(f"[ERROR] Generation failed: {result}")
        return False

    generate_uuid = result.get("data", {}).get("generateUuid")
    if not generate_uuid:
        print(f"[ERROR] No generateUuid in response")
        return False

    print(f"  Task UUID: {generate_uuid}")

    # 轮询状态
    max_attempts = 60
    for attempt in range(max_attempts):
        time.sleep(5)  # 每5秒检查一次

        status_result = check_status(generate_uuid)
        status = status_result.get("data", {}).get("status")

        print(f"  Status check {attempt + 1}/{max_attempts}: {status}")

        if status == "SUCCESS":
            # 获取图片URL
            images = status_result.get("data", {}).get("images", [])
            if images:
                image_url = images[0].get("imageUrl")
                if image_url:
                    return download_image(image_url, filename)
            print(f"[ERROR] No image URL in response")
            return False
        elif status == "FAILED":
            print(f"[ERROR] Generation failed")
            return False

    print(f"[ERROR] Timeout waiting for generation")
    return False

def main():
    """主函数 - 生成所有待处理的资源"""

    # 确保输出目录存在
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    # 读取任务文件
    task_file = r"E:\AI_Project\MS\.task\art_generation_task.json"
    with open(task_file, 'r', encoding='utf-8') as f:
        task_data = json.load(f)

    requests_list = task_data.get("generation_requests", [])

    print(f"[INFO] Found {len(requests_list)} generation requests")
    print(f"[INFO] Output directory: {OUTPUT_DIR}")

    success_count = 0
    fail_count = 0

    for req in requests_list:
        if req.get("status") == "completed":
            print(f"\n[SKIP] Already completed: {req['filename']}")
            continue

        success = generate_image(
            prompt=req["prompt"],
            negative_prompt=req.get("negative_prompt", ""),
            width=req["width"],
            height=req["height"],
            filename=req["filename"]
        )

        if success:
            req["status"] = "completed"
            success_count += 1
        else:
            req["status"] = "failed"
            fail_count += 1

        # 更新任务文件
        with open(task_file, 'w', encoding='utf-8') as f:
            json.dump(task_data, f, indent=2, ensure_ascii=False)

    print(f"\n[DONE] Success: {success_count}, Failed: {fail_count}")

if __name__ == "__main__":
    main()
